datafusion/physical-expr/src/expressions/case/literal_lookup_table/mod.rs

```diff
@@ -0,0 +1,440 @@
+            else {
+                return None;
+            };
+
+            null_scalar
+        };
+
+        {
+            let data_type = when_literals[0].data_type();
+
+            // If not all the WHEN literals are the same data type we cannot use this optimization
+            if when_literals.iter().any(|l| l.data_type() != data_type) {
+                return None;
+            }
+        }
+
+        {
+            let data_type = then_literals[0].data_type();
+
+            // If not all the then and the else literals are the same data type we cannot use this optimization
+            if then_literals.iter().any(|l| l.data_type() != data_type) {
+                return None;
+            }
+
+            if else_expr.data_type() != data_type {
+                return None;
+            }
+        }
+
+        let output_array = ScalarValue::iter_to_array(
+            then_literals
+                .iter()
+                // The else is in the end
+                .chain(std::iter::once(&else_expr))
+                .cloned(),
+        )
+        .ok()?;
+
+        let lookup = try_creating_lookup_table(
+            when_literals,
+            // The else expression is in the end
+            output_array.len() as i32 - 1,
+        )
+        .ok()?;
+
+        Some(Self {
+            lookup,
+            values_to_take_from: output_array,
+        })
+    }
+
+    pub(in super::super) fn create_output(
+        &self,
+        expr_array: &ArrayRef,
+    ) -> datafusion_common::Result<ArrayRef> {
+        let take_indices = self.lookup.match_values(expr_array)?;
+
+        // Zero-copy conversion
+        let take_indices = Int32Array::from(take_indices);
+
+        // An optimize version would depend on the type of the values_to_take_from
+        // For example, if the type is view we can just keep pointing to the same value (similar to dictionary)
+        // if the type is dictionary we can just use the indices as is (or cast them to the key type) and create a new dictionary array
+        let output = arrow::compute::take(&self.values_to_take_from, &take_indices, None)
+            .map_err(|e| arrow_datafusion_err!(e))?;
+
+        Ok(output)
+    }
+}
+
+/// Lookup table for mapping literal values to their corresponding indices in the THEN clauses
+///
+/// The else index is used when a value is not found in the lookup table
+pub(super) trait WhenLiteralIndexMap: Debug + Send + Sync {
+    /// Try creating a new lookup table from the given literals and else index
+    ///
+    /// `literals` are guaranteed to be unique and non-nullable
+    fn try_new(
+        unique_non_null_literals: Vec<ScalarValue>,
+        else_index: i32,
+    ) -> datafusion_common::Result<Self>
+    where
+        Self: Sized;
+
+    /// Return indices to take from the literals based on the values in the given array
+    fn match_values(&self, array: &ArrayRef) -> datafusion_common::Result<Vec<i32>>;
```

Comment:
Maybe `map_to_value_indices` or something along those lines would make this more self explanatory?